### PR TITLE
[PT FE] Support aten::take operation

### PR DESCRIPTION
**Pull request by vijaykr338**
### Details:
I was able to implement aten::take and aten::Delete function,

but aten::str posed really high difficult, i was only able to partially implement the it 

currently, my implementation works for constant tensors, but for dynamic tensors, it currently stores shape and dtype for later use in debugging.
 
I have another question, how do you write tests for functions which are not implemented in Aten library yet? 

I was able to write test cases for aten::take by following examples existing pytest files in the PyTorch layer tests folder.  

but when I did the same for Delete and str, they consistently gave error

this is my first contribution to openVINO, I would greatly appreciate any guidance or assistance.
  

### Tickets:
 - #28707

### CODE REVIEW

**Comment by mvafin** [comment]
Please fix code style and add tests for aten::str and aten::Delete.

**Comment by vijaykr338** [comment]
> Please fix code style and add tests for aten::str and aten::Delete.

Sir @mvafin,  I will fix the code style but 

I am having issues writing tests for str  

```
E       openvino._pyopenvino.OpConversionFailure: Check 'is_conversion_successful' failed at src/frontends/pytorch/src/frontend.cpp:174:
E       FrontEnd API failed with OpConversionFailure:
E       Model wasn't fully converted.
E       Summary:
E       -- No conversion rule found for operations: aten::str
```
I have added the aten::str in the op_table.cpp, along with the function name

**Comment by vijaykr338** [comment]
> > > Please fix code style and add tests for aten::str and aten::Delete.
> > 
> > 
> > Sir @mvafin, I will fix the code style but
> > I am having issues writing tests for str
> > ```
> > E       openvino._pyopenvino.OpConversionFailure: Check 'is_conversion_successful' failed at src/frontends/pytorch/src/frontend.cpp:174:
> > E       FrontEnd API failed with OpConversionFailure:
> > E       Model wasn't fully converted.
> > E       Summary:
> > E       -- No conversion rule found for operations: aten::str
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I have added the aten::str in the op_table.cpp, along with the function name
> 
> Did you build openvino? It seems you are testing the release version of openvino, not the one you built.

I have recognized the issue 
There was an issue with the registeration of aten::str in op_table.cpp, I am working on it now

**Comment by vijaykr338** [comment]
@mvafin I have updated the implementation and test file for aten::take and aten::delete 
since there is no delete function in torch library, I assumed that it is meant to remove an element from input tensor, by slicing the portion before and after that element and then concatenating those slices 

so my model looks like this for aten::delete 
```
 def create_model(self):
        class aten_delete(torch.nn.Module):
            def forward(self, x, index):
                idx = int(index.item())
                # Delete along axis 0 by concatenating slices before and after idx.
                return torch.cat([x[:idx], x[idx+1:]], dim=0)
        ref_net = None
        return aten_delete(), ref_net, ["aten::slice", "aten::cat"]
```

please check

**Comment by vijaykr338** [comment]
for aten::Str, I'd like to confirm it's intended function 

should it mimic python's inbuilt str() function? the general function is to convert tensors into string representation

my current implementation does not work, as it gives errors 
```
E       openvino._pyopenvino.OpConversionFailure: Check 'is_conversion_successful' failed at src/frontends/pytorch/src/frontend.cpp:174:
E       FrontEnd API failed with OpConversionFailure:
E       Model wasn't fully converted. Failed operations detailed log:
E       -- prim::Constant with a message:
E       String constant cannot be converted to OpenVINO opset and should be removed by consuming operation.
E       Summary:
E       -- Conversion is failed for: prim::Constant
```
according to docs,  raw string constants aren’t directly supported in the OpenVINO opset - instead I found StringTensorUnpack - which can make break the string into three parts to make it usable for further operations

**Comment by vijaykr338** [comment]
@mvafin what do you think?

**Comment by mvafin** [comment]
@vijaykr338 Openvino support string tensors if created inside the model. However we cannot return or accept strings as model inputs/outputs. What model are you testing this on? I think it might be very hard or even impossible to make such a model that will be using aten::str inside.

As for delete, you are clearly not using aten::Delete inside the model code. I suspect that this operation is `del` that removes an element from `dict`. If so you wouldn't be able to implement since openvino doesn't support dictionaries, we decompose them.

**Comment by vijaykr338** [comment]
> @vijaykr338 Openvino support string tensors if created inside the model. However we cannot return or accept strings as model inputs/outputs. What model are you testing this on? I think it might be very hard or even impossible to make such a model that will be using aten::str inside.
> 
> As for delete, you are clearly not using aten::Delete inside the model code. I suspect that this operation is `del` that removes an element from `dict`. If so you wouldn't be able to implement since openvino doesn't support dictionaries, we decompose them.

given such constraints, I think it would be wise to drop these 2 operations and focus on refining aten::Take instead 

for aten::str, I thought about using an encoding-decoding pipeline to convert the string data to numerical data then back to it which is basically openVINO tokenizers, but that simply wont work. 

similarly for aten::delete, I had another plan of making a 'mask' tensor with true and false values to filter out false indice elements from target tensor, but that wouldnt work either considering the constraints 

therefore I think it's better for me to refine just the take function, what do you think?

**Comment by vijaykr338** [comment]
@mvafin could you please take a look at my take implementation and test file?

**Comment by vijaykr338** [comment]
Hey @rkazants @mvafin! Just wanted to follow up on this PR — totally understand it’s a busy time with GSOC around the corner. When you get a chance, I’d really appreciate any feedback so I can make improvements. Thanks in advance!

**Comment by vijaykr338** [comment]
@mvafin the tests for pytorch ran fine on the other operating systems, I am not sure why Ubuntu 22.04 failed, even the logs didn't help

**Comment by vijaykr338** [comment]
@mvafin the tests for pytorch ran fine on the other operating systems, I am not sure why Ubuntu 22.04 failed, even the logs didn't help

**Comment by vijaykr338** [comment]
@mvafin @rkazants  the tests for pytorch ran fine on the other operating systems, I am not sure why Ubuntu 22.04 failed, even the logs didn't help

**Comment by vijaykr338** [comment]
@mvafin @rkazants  I will explain my strategy for the take function 

implementation -> 
We receive the input tensor and a set of indices -> flatten the tensor into a 1D list -> pick out the elements from at the specified indices -> gather them and then return 

checks ->
we are checking if the input tensor has atleast one dimension
and also an out of bounds check IF the tensor's shape is static, otherwise we skip this check 

what do you think?

File: tests/layer_tests/pytorch_tests/test_take.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+                return torch.take(x, indices)
+
+        ref_net = None
+        return aten_take(), ref_net, "aten::take"
+
+    @pytest.mark.nightly
+    @pytest.mark.precommit
+    @pytest.mark.precommit_torch_export
+    @pytest.mark.parametrize("input_shape", [(10,), (3, 4), (2, 3, 4), (100,)])
+    @pytest.mark.parametrize("indices_shape", [(5,), (2, 2), (3, 2), (50,)])
```
Comment (mvafin) [inline tests/layer_tests/pytorch_tests/test_take.py:27]:
```suggestion
    @pytest.mark.parametrize("input_shape", [(10,), (3, 4), (2, 3, 4)])
    @pytest.mark.parametrize("indices_shape", [(5,), (2, 2), (3, 2)])
```
Lets reduce number of tests

**Comment by mvafin** [comment]
> @mvafin @rkazants the tests for pytorch ran fine on the other operating systems, I am not sure why Ubuntu 22.04 failed, even the logs didn't help

It failed by timeout. Tests either too long or we just have added too many new tests

**Comment by vijaykr338** [comment]
> > @mvafin @rkazants the tests for pytorch ran fine on the other operating systems, I am not sure why Ubuntu 22.04 failed, even the logs didn't help
> 
> It failed by timeout. Tests either too long or we just have added too many new tests

I see

**Comment by vijaykr338** [comment]
@mvafin could you please re-rerun the CI/CD tests?

**Comment by vijaykr338** [comment]
@mvafin can you check if everything is alright?

**Comment by vijaykr338** [comment]
@mvafin @rkazants Just checking in — is there anything else I should change or address in this PR? I noticed that the CI failures don’t seem related to my implementation. 

I would really appreciate it if you could take a look whenever you get a chance

thanks

**Comment by mlukasze** [comment]
build_jenkins

**Comment by vijaykr338** [comment]
@mlukasze the tests have completed, could you take a look?

**Comment by mlukasze** [comment]
build_jenkins

**Comment by vijaykr338** [comment]
@mlukasze apologies for mentioning again, I believe that this PR can be merged now since all checks have passed successfully.

File: src/frontends/pytorch/src/op/take.cpp [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    auto indices = context.get_input(1);
+    auto input_shape = input.get_partial_shape();
+    if (input_shape.rank().is_static() && input_shape.rank().get_length() == 0) {
+        FRONT_END_OP_CONVERSION_CHECK(false, "input tensor MUST be non-scalar");
+    }
+    auto new_shape = context.mark_node(v0::Constant::create(element::i64, Shape{1}, {-1}));
+    input = context.mark_node(std::make_shared<v1::Reshape>(input, new_shape, false));
+    indices = context.mark_node(std::make_shared<v0::Convert>(indices, element::i64));
+    auto input_size = context.mark_node(std::make_shared<v3::ShapeOf>(input, element::i64));
+    indices = normalize_axis(context, indices, input_size);
```
Comment (mmikolajcz) [inline src/frontends/pytorch/src/op/take.cpp]:
```suggestion
```
This may not be needed since v8::Gather supports negative indices

File: src/frontends/pytorch/src/op/take.cpp [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        auto indices_vals = indices_const->cast_vector<int64_t>();
+        for (const auto& idx : indices_vals) {
+            FRONT_END_OP_CONVERSION_CHECK(idx >= 0 && idx < flat_size,
+                                          "Index out of bounds: ",
+                                          idx,
+                                          ". Valid range: [0, ",
+                                          flat_size,
+                                          ").");
+        }
+    }
```
Comment (mmikolajcz) [inline src/frontends/pytorch/src/op/take.cpp]:
```suggestion
```
Not sure whether this would be needed, usually bounds are already verified on PyTorch side

File: src/frontends/pytorch/src/op/take.cpp [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+using namespace ov::op;
+
+OutputVector translate_take_op(const NodeContext& context) {
+    num_inputs_check(context, 2, 2, true);
+    auto input = context.get_input(0);
+    auto indices = context.get_input(1);
+    auto input_shape = input.get_partial_shape();
+    if (input_shape.rank().is_static() && input_shape.rank().get_length() == 0) {
+        FRONT_END_OP_CONVERSION_CHECK(false, "input tensor MUST be non-scalar");
+    }
```
Comment (mmikolajcz) [inline src/frontends/pytorch/src/op/take.cpp:27]:
```suggestion
```
Looking on how torch.take functions, scalar tensors should be allowed, for example:
`torch.take(torch.tensor(5), torch.tensor(0)) -> tensor(5)`
`torch.take(torch.tensor(5), torch.tensor([0])) -> tensor([5])`

Is there any reason for such check? I've modified test suite a bit and it seem that scalars should work correctly.

File: src/frontends/pytorch/src/op/take.cpp [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+            FRONT_END_OP_CONVERSION_CHECK(idx >= 0 && idx < flat_size,
+                                          "Index out of bounds: ",
+                                          idx,
+                                          ". Valid range: [0, ",
+                                          flat_size,
+                                          ").");
+        }
+    }
+    auto axis_constant = context.mark_node(v0::Constant::create(element::i64, Shape{}, {0}));
+    auto gather = context.mark_node(std::make_shared<v8::Gather>(input, indices, axis_constant));
```
Comment (mmikolajcz) [inline src/frontends/pytorch/src/op/take.cpp:32]:
```suggestion
    auto gather = context.mark_node(std::make_shared<v8::Gather>(input, indices, axis_constant));
    if (!context.input_is_none(2)) {
        context.mutate_input(2, gather);
    }
```
aten::take supports out input at index 2 that needs to be handled

File: src/frontends/pytorch/src/op/take.cpp [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+namespace ov {
+namespace frontend {
+namespace pytorch {
+namespace op {
+
+using namespace ov::op;
+
+OutputVector translate_take_op(const NodeContext& context) {
+    num_inputs_check(context, 2, 2, true);
```
Comment (mmikolajcz) [inline src/frontends/pytorch/src/op/take.cpp]:
```suggestion
    num_inputs_check(context, 2, 3);
```
Optional out input should be allowed.
I'm not sure about `allow_complex=true`, cases like that require some special handling and additional test cases for complex types.

File: tests/layer_tests/pytorch_tests/test_take.py [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+
+        ref_net = None
+        return aten_take(), ref_net, "aten::take"
+
+    @pytest.mark.nightly
+    @pytest.mark.precommit
+    @pytest.mark.precommit_torch_export
+    @pytest.mark.parametrize("input_shape", [(10,), (3, 4), (2, 3, 4)])
+    @pytest.mark.parametrize("indices_shape", [(5,), (2, 2), (3, 2)])
+    def test_take(self, input_shape, indices_shape, ie_device, precision, ir_version):
```
Comment (mmikolajcz) [inline tests/layer_tests/pytorch_tests/test_take.py:28]: Please add test case for scenario with out. This would require few changes for whole class, you can refer to similar op: https://github.com/openvinotoolkit/openvino/blob/master/tests/layer_tests/pytorch_tests/test_take_along_dim.py#L51
